logstash/parallel_process.go

```diff
@@ -0,0 +1,277 @@
+	}
+
+	logFile, err := newDeletedTempFile("", "")
+	if err != nil {
+		CleanupTestStreams(testStream)
+		return nil, err
+	}
+
+	configDir, err := getConfigFileDir(configs)
+	if err != nil {
+		CleanupTestStreams(testStream)
+		_ = logFile.Close()
+		return nil, err
+	}
+
+	args := []string{
+		"-w", // Make messages arrive in order.
+		"1",
+		"--debug",
+		"-e",
+		fmt.Sprintf(
+			"input { %s } "+
+				"output { %s }",
+			strings.Join(logstashInput, " "), strings.Join(logstashOutput, " ")),
+		"--config",
+		configDir,
+		"--log",
+		logFile.Name(),
+	}
+
+	p, err := newParallelProcessWithArgs(logstashPath, args, getLimitedEnvironment(os.Environ(), keptEnvVars))
+	if err != nil {
+		CleanupTestStreams(testStream)
+		_ = logFile.Close()
+	}
+	p.configDir = &configDir
+	p.log = logFile
+	p.streams = testStream
+	return p, nil
+}
+
+// newProcessWithArgs performs the non-Logstash specific low-level
+// actions of preparing to spawn a child process, making it easier to
+// test the code in this package.
+func newParallelProcessWithArgs(command string, args []string, env []string) (*ParallelProcess, error) {
+	c := exec.Command(command, args...)
+	c.Env = env
+
+	// Save the process's stdout and stderr since an early startup
+	// failure (e.g. JVM issues) will get dumped there and not in
+	// the log file.
+	var b bytes.Buffer
+	c.Stdout = &b
+	c.Stderr = &b
+
+	return &ParallelProcess{
+		child: c,
+		stdio: &b,
+	}, nil
+}
+
+// Start starts a Logstash child process with the previously supplied
+// configuration.
+func (p *ParallelProcess) Start() error {
+	log.Info("Starting %q with args %q.", p.child.Path, p.child.Args[1:])
+	return p.child.Start()
+}
+
+// Wait blocks until the started Logstash process terminates and
+// returns the result of the execution.
+func (p *ParallelProcess) Wait() (*ParallelResult, error) {
+	if p.child.Process == nil {
+		return nil, errors.New("Can't wait on an unborn process.")
+	}
+	log.Debug("Waiting for child with pid %d to terminate.", p.child.Process.Pid)
+
+	waiterr := p.child.Wait()
+
+	// Save the log output regardless of whether the child process
+	// succeeded or not.
+	logbuf, logerr := ioutil.ReadAll(p.log)
+	if logerr != nil {
+		// Log this weird error condition but don't let it
+		// fail the function. We don't care about the log
+		// contents unless Logstash fails, in which we'll
+		// report that problem anyway.
+		log.Error("Error reading the Logstash logfile: %s", logerr.Error())
+	}
+	outbuf, _ := ioutil.ReadAll(p.stdio)
+
+	result := ParallelResult{
+		Events:  [][]Event{},
+		Log:     string(logbuf),
+		Output:  string(outbuf),
+		Success: waiterr == nil,
+	}
+	if waiterr != nil {
+		if strings.Contains(result.Log, `:message=>"An unexpected error occurred!", :error=>"closed stream", :class=>"IOError"`) {
+			log.Warning("Workaround for IOError in unix.rb on stop, process result anyway. (see https://github.com/logstash-plugins/logstash-input-unix/pull/18)")
```

Comment:
You are right, this error message check was too specific to the version of Logstash, which I use at the moment (2.4.1). 

I updated the PR with a regexp for this error message check. I tested the following versions (latest patch release of every minor):

Successful, warning is shown:

* 2.4.1
* 2.3.4

Successful, interestingly no warning is show (behavior of Logstash is different in these versions, logstash-input-unix did not change to much):

* 2.2.4 
* 2.1.3
* 2.0.0

For the unix domain socket feature to work (especially to shutdown the Logstash pipeline after the unix domain socket is closed), at least version 2.0.2 (or commit https://github.com/logstash-plugins/logstash-input-unix/commit/5981b148909ed6da5a4b8365bb466fe65d69d0d2) of logstash-input-unix is required. Therefore it does not work with Logstash prior to version 2.x. I guess it is ok to only support Logstash version > 2.x with this feature, because this feature is anyways limited to *nix systems and the variant with stdin is always available as a backup.

As far as I understand, logstash-filter-verifier does not yet support Logstash 5.x (#8), so I skipped these for now.